src/mono/wasm/runtime/binding_support.js

```diff
@@ -162,49 +160,174 @@ var BindingSupportLib = {
 			this._js_owned_object_registry = new FinalizationRegistry(this._js_owned_object_finalized.bind(this));
 		},
 
-		_get_weak_delegate_from_handle: function (id) {
-			var result = null;
+		_js_owned_object_finalized: function (gc_handle) {
+			// The JS function associated with this gc_handle has been collected by the JS GC.
+			// As such, it's not possible for this gc_handle to be invoked by JS anymore, so
+			//  we can release the tracking weakref (it's null now, by definition),
+			//  and tell the C# side to stop holding a reference to the managed delegate.
+			this._js_owned_object_table.delete(gc_handle);
+			this.release_js_owned_object_by_handle(gc_handle);
+		},
+
+		_lookup_js_owned_object: function (gc_handle) {
+			if (!gc_handle)
+				return null;
+			var wr = this._js_owned_object_table.get(gc_handle);
+			if (wr) {
+				return wr.deref();
+				// TODO: could this be null before _js_owned_object_finalized was called ?
+				// TODO: are there race condition consequences ?
+			}
+			return null;
+		},
+
+		_wrap_js_thenable_as_task: function (thenable) {
+			this.bindings_lazy_init ();
+			if (!thenable)
+				return null;
+
+			// hold strong JS reference to thenable while in flight
+			// ideally, this should be hold alive by lifespan of the resulting C# Task, but this is good cheap aproximation
+			var thenable_js_handle = this.mono_wasm_free_list.length ? this.mono_wasm_free_list.pop() : this.mono_wasm_ref_counter++;
+			this.mono_wasm_object_registry[thenable_js_handle] = thenable;
+
+			// TODO optimization: return the tcs.Task on the same call
+			// note that we do not implement promise/task roundtrip
```

Comment:
maybe be more specific in this comment, i.e. "... do not implement promise/task round trip, which means passing a Task into JS and then back out will create a new Task with a different type"

Reply:
We are converting to another type because we must. I have different point here.
But we could still recover original instance from it when it would be passed back.
I improved both comments